pkg/healthcheck/healthcheck.go

```diff
@@ -2191,6 +2209,90 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
 	return nil
 }
 
+// Check if there's a pod with the opaque-ports annotation defined but a
+// service selecting the aforementioned pod doesn't define it
+func (hc *HealthChecker) checkMisconfiguredOpaquePortAnnotations(ctx context.Context, namespace string) error {
+	services, err := hc.GetServices(ctx)
+	if err != nil {
+		return err
+	}
+
+	pods, err := hc.kubeAPI.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	var errStrings []string
+
+	for _, service := range services {
```

Comment:
one alternative here that would help us get rid of the `serviceMatchesPod()` fn is to use the Kubernetes client and fetch pods for each service based on the service's selector.

 ```suggestion
	for _, service := range services {
	     if service.spec.ClusterIP == "None" {
	       continue // skip headless
	      }
	      
	     labelSet := labels.Set(service.Spec.Selector)
	     pods, err := hc.kubeAPI.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions { LabelSelector: labelSet.AsSelector()}
	     if err != nil {
	         return err
	      }
	     if mistmatch := (....)
	  }
```

This has a few advantages:
1. We can get rid of the pods and error check on lines 2220,2221.
2. We can get rid of the `serviceMatchesPod` function
3. Simplify the logic a little bit.

The downside is we don't have a client-go informer so we would make as many API requests to the Kubernetes server as there are services in that namespace. This might not be a problem though since `check` is meant to be used sparingly, I doubt this would bring anyone's K8s control plane down (or flood it with requests).

wdyt @migue ? I'm not particular to any one way of doing it, think the way it is now works fine -- just an alternative.

Reply:
That was my first option but wanted to minimize the number of requests to the control plane. If you folks think this is not going to be a problem I think we should go forward with your proposal.

I can just talk for my current setup; we have ~50 services so this would be <50 requests (we have a bunch of headless services) which is not a big deal, but not sure about the number of services in general (I guess people won't have millions of services in a single cluster :D )

Reply:
Let's put a hold on this for a bit and see what the second reviewer thinks perhaps. I got bitten by this before so I'm also in two minds about it.

Reply:
Sounds good!!

Again, thanks a lot reviewing the pull!!

Reply:
So I lean towards @mateiidavid's suggestion on this and going with the selector approach. This could lead to a lot of kube API calls depending on the number of services, but the `check --proxy` command is not in a critical path and we can simplify some of the logic here. If it is a noticeable issue then we can optimize for that down the line.

Reply:
I am okay with it; as I said, it was my first option but wasn't sure about the potential "performance" implications ...

Will update the pull accordingly